Objects/frameobject.c

```diff
@@ -1707,70 +1776,92 @@ frame_setlineno(PyObject *op, PyObject* p_new_lineno, void *Py_UNUSED(closure))
     while (start_stack > best_stack) {
         if (top_of_stack(start_stack) == Except) {
             /* Pop exception stack as well as the evaluation stack */
-            PyObject *exc = PyStackRef_AsPyObjectBorrow(_PyFrame_StackPop(f->f_frame));
+            PyObject *exc = PyStackRef_AsPyObjectBorrow(_PyFrame_StackPop(self->f_frame));
             assert(PyExceptionInstance_Check(exc) || exc == Py_None);
             PyThreadState *tstate = _PyThreadState_GET();
             Py_XSETREF(tstate->exc_info->exc_value, exc == Py_None ? NULL : exc);
         }
         else {
-            PyStackRef_XCLOSE(_PyFrame_StackPop(f->f_frame));
+            PyStackRef_XCLOSE(_PyFrame_StackPop(self->f_frame));
         }
         start_stack = pop_value(start_stack);
     }
     /* Finally set the new lasti and return OK. */
-    f->f_lineno = 0;
-    f->f_frame->instr_ptr = _PyFrame_GetBytecode(f->f_frame) + best_addr;
+    self->f_lineno = 0;
+    self->f_frame->instr_ptr = _PyFrame_GetBytecode(self->f_frame) + best_addr;
     return 0;
 }
 
+/*[clinic input]
+@getter
+frame.f_trace as frame_trace
+
+Return the trace function for this frame, or None if no trace function is set.
+[clinic start generated code]*/
+
 static PyObject *
-frame_gettrace(PyObject *op, void *Py_UNUSED(closure))
+frame_trace_get_impl(PyFrameObject *self)
+/*[clinic end generated code: output=5475cbfce07826cd input=7b47c6f37b8ad405]*/
 {
-    PyFrameObject *f = PyFrameObject_CAST(op);
-    PyObject* trace = f->f_trace;
-    if (trace == NULL)
+    PyObject* trace = self->f_trace;
+    if (trace == NULL) {
         trace = Py_None;
+    }
     return Py_NewRef(trace);
 }
 
+/*[clinic input]
+@setter
+frame.f_trace as frame_trace
```

Comment:
```suggestion
@critical_section
@setter
frame.f_trace as frame_trace
```

I think this should also use critical section as it mutates the `self->f_trace` field.